From 1eff4199b98af1b542ea414db5be63c61b55e245 Mon Sep 17 00:00:00 2001
From: Fabrizio Furano <furano@cern.ch>
Date: Wed, 4 Oct 2017 15:18:57 +0200
Subject: [PATCH 4/4] Fix for Github #596 Postprocessing a kxr_close should
 finish gracefully the request (fix for the double sending) In any case, make
 sure that any buffer is not sent twice Some indentation fixes

---
 src/XrdHttp/XrdHttpReq.cc | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/src/XrdHttp/XrdHttpReq.cc b/src/XrdHttp/XrdHttpReq.cc
index a172264..af88962 100644
--- a/src/XrdHttp/XrdHttpReq.cc
+++ b/src/XrdHttp/XrdHttpReq.cc
@@ -1752,6 +1752,9 @@ int XrdHttpReq::PostProcessHTTPReq(bool final_) {
           default: //read or readv
           {
             
+            // Nothing to do if we are postprocessing a close
+            if (ntohs(xrdreq.header.requestid) == kXR_close) return 1;
+            
 	    // Close() if this was the third state of a readv, otherwise read the next chunk
 	    if ((reqstate == 3) && (ntohs(xrdreq.header.requestid) == kXR_readv)) return 1;
 	    
@@ -1812,17 +1815,21 @@ int XrdHttpReq::PostProcessHTTPReq(bool final_) {
 		writtenbytes += iovP[i].iov_len;
               }
               
+            // Let's make sure that we avoid sending the same data twice,
+            // in the case where PostProcessHTTPReq is invoked again
+            this->iovN = 0;
+            
             return 0;
           }
 
-        }
+        } // switch reqstate
 
 
       }
 
 
       break;
-    }
+    } // case GET
 
 
     case XrdHttpReq::rtPUT:
-- 
2.7.5

